File: task2/d_calculator.py
#
#   Примечание: данное решение не зашло (неправильный ответ на 2 тесте, хз почему)
#   Смотри решение на плюсах
#
max_number = int(input().strip())

number_of_operations = [0 for i in range(max_number + 1)]
previous_number = [-1 for i in range(max_number + 1)]

for i in range(2, max_number + 1):
    possible_previous_numbers = [i - 1]
    if i % 2 == 0:
        possible_previous_numbers.append(i // 2)
    if i % 3 == 0:
        possible_previous_numbers.append(i // 3)

    possible_previous_operations = [(number_of_operations[prev], prev) 
                                    for prev in possible_previous_numbers]
    n_operations, prev_number = min(possible_previous_operations)
    number_of_operations[i] = n_operations + 1
    previous_number[i] = prev_number

print(number_of_operations[max_number])

result_path = []
curr_number = max_number
while curr_number >= 0:
    result_path.append(curr_number)
    curr_number = previous_number[curr_number]
print(" ".join([str(x) for x in reversed(result_path)]))

File: task2/d_calculator.cpp
#include <iostream>
#include <vector>
#include <string>

int main() {
    int max_number;
    std::cin >> max_number;
    std::vector<int> number2operations(max_number + 1);
    std::vector<int> number2prev_number(max_number + 1);
    number2operations[1] = 0;
    number2prev_number[1] = -1;

    for (int i = 2; i <= max_number; ++i) {
        int prev_number = i - 1;
        int prev_operations = number2operations[prev_number];
        if (i % 2 == 0 && number2operations[i / 2] < prev_operations) {
            prev_number = i / 2;
            prev_operations = number2operations[prev_number];
        }
        if (i % 3 == 0 && number2operations[i / 3] < prev_operations) {
            prev_number = i / 3;
            prev_operations = number2operations[prev_number];
        }
        number2operations[i] = prev_operations + 1;
        number2prev_number[i] = prev_number;
    }

    std::cout << number2operations[max_number] << std::endl;
    int curr_number = max_number;
    std::string path = "";
    while (curr_number > 0) {
        path = std::to_string(curr_number) + " " + path;
        curr_number = number2prev_number[curr_number];
    }
    std::cout << path << std::endl;

    return 0;
}